common/network-shuffle/src/main/java/org/apache/spark/network/shuffle/RemoteBlockPushResolver.java

```diff
@@ -410,17 +498,41 @@ public MergeStatuses finalizeShuffleMerge(FinalizeShuffleMerge msg) throws IOExc
           + "with the current attempt id %s stored in shuffle service for application %s",
           msg.appAttemptId, appShuffleInfo.attemptId, msg.appId));
     }
-    Map<Integer, AppShufflePartitionInfo> shufflePartitions =
-      appShuffleInfo.partitions.remove(msg.shuffleId);
+    AtomicReference<Map<Integer, AppShufflePartitionInfo>> shuffleMergePartitionsRef =
+      new AtomicReference<>(null);
+    // Metadata of the determinate stage shuffle can be safely removed as part of finalizing
+    // shuffle merge. Currently once the shuffle is finalized for a determinate stages, retry
+    // stages of the same shuffle will have shuffle push disabled.
+    if (msg.shuffleMergeId == DETERMINATE_SHUFFLE_MERGE_ID) {
+      shuffleMergePartitionsRef.set(
+        appShuffleInfo.partitions.remove(msg.shuffleId).shuffleMergePartitions);
```

Comment:
`appShuffleInfo.partitions.remove(msg.shuffleId)` can be `null` - check before setting

Reply:
Just curious, something like a duplicate finalize request can return null?

Reply:
You are right, with the current state of the code, this should not cause. I would prefer to be more defensive in this case though, given the various corner cases we have and any change can result in this.